query/src/storages/fuse/io/block_reader.rs

```diff
@@ -15,121 +15,119 @@
 
+#[derive(Clone)]
 pub struct BlockReader {
-    data_accessor: Operator,
-    path: String,
-    block_schema: DataSchemaRef,
-    arrow_table_schema: ArrowSchema,
-    projection: Vec<usize>,
-    file_len: u64,
-    metadata_reader: BlockMetaReader,
+    operator: Operator,
+    arrow_schema: Arc<Schema>,
+    projected_schema: DataSchemaRef,
+    parquet_schema_descriptor: SchemaDescriptor,
 }
 
 impl BlockReader {
-    pub fn new(
-        data_accessor: Operator,
-        path: String,
-        table_schema: DataSchemaRef,
+    pub fn create(
+        operator: Operator,
+        schema: DataSchemaRef,
         projection: Vec<usize>,
-        file_len: u64,
-        reader: BlockMetaReader,
-    ) -> Self {
-        let block_schema = Arc::new(table_schema.project(projection.clone()));
-        let arrow_table_schema = table_schema.to_arrow();
-        Self {
-            data_accessor,
-            path,
-            block_schema,
-            arrow_table_schema,
-            projection,
-            file_len,
-            metadata_reader: reader,
+    ) -> Result<Arc<BlockReader>> {
+        let projected_schema = DataSchemaRef::new(schema.project(projection));
+
+        let arrow_schema = schema.to_arrow();
+        let parquet_schema_descriptor = to_parquet_schema(&arrow_schema)?;
+        Ok(Arc::new(BlockReader {
+            operator,
+            projected_schema,
+            parquet_schema_descriptor,
+            arrow_schema: Arc::new(arrow_schema),
+        }))
+    }
+
+    fn to_deserialize(
+        meta: &ColumnMeta,
+        chunk: Vec<u8>,
+        rows: usize,
+        descriptor: &ColumnDescriptor,
+        field: Field,
+    ) -> Result<ArrayIter<'static>> {
+        let pages = PageIterator::new(
+            std::io::Cursor::new(chunk),
+            meta.num_values as i64,
+            Compression::Lz4,
+            descriptor.clone(),
+            Arc::new(|_, _| true),
+            vec![],
+        );
+
+        let descriptor_type = descriptor.type_();
+        let decompressor = BasicDecompressor::new(pages, vec![]);
+        Ok(column_iter_to_arrays(
+            vec![decompressor],
+            vec![descriptor_type],
+            field,
+            rows,
+        )?)
+    }
+
+    async fn read_columns(self, part: PartInfoPtr) -> Result<(usize, Vec<ArrayIter<'static>>)> {
+        let part = FusePartInfo::from_part(&part)?;
+
+        let rows = part.nums_rows;
+        // TODO: add prefetch column data.
+        let mut columns_array_iter = Vec::with_capacity(0);
+        for (index, column_meta) in &part.columns_meta {
+            let mut column_reader = self
+                .operator
+                .object(&part.location)
+                .range_reader(column_meta.offset, column_meta.length);
+
+            let mut column_chunk = vec![0; column_meta.length as usize];
+            // TODO: check thread name
+            column_reader.read_exact(&mut column_chunk).await?;
```

Comment:
does it mean that columns of a given block will be fetched sequentially?

Reply:
I will solve this in the next pr.